### ISSUE TITLE
History: add a GUI viewer

### ISSUE THREAD

**Issue by amahabal** [amahabal/PySeqsee#35]
Currently, there is a way to inspect history. When the --history flag is passed in, we drop into a python interpreter that can show a few things.

It would be nice to have a UI for that, would make it quicker to inspect things.

**Comment by superMDguy** [comment]
I'm going to be implementing this within `core/history.py`, is that what you were thinking?

**Comment by amahabal** [comment]
Whether it should be in that file or a separate one should depend on how
big it is. I suggest doing it in that file, and we can move it later if
that file is getting very bloated...

Thanks!
  Abhijit

**Comment by superMDguy** [comment]
It's working as of the latest commit.  I currently just have it putting the text that was printed into the terminal into the GUI window, but I think we should change that.  Any ideas on ways to make it look better?

**Comment by amahabal** [comment]
I tried running it, and could not see the output. I used the following command-line:

 farg_tool run seqsee --sequence 1 2 3 4 5 --history

With this, the main UI starts; when it finishes, a window flashes on to the screen (the history window, I take it), and almost immediately it disappears and the app exits. Are you using a different invocation?

**Comment by superMDguy** [comment]
Looks like it's working when I click the "quit" button (which is what I was doing), but not when it quits by itself when the answer is found.  I'm currently looking into this.

**Comment by amahabal** [comment]
I see, yes, I was able to see it.

What do you think of these two directions?
(A) The history window can co-exist with the mail window, so that one can explore things mid-way through a run. Hopefully, that should be a simple change.
(B) Make things navigable via clicking (instead of typing the HID in, for instance). As an example of this, you can see how the coderack view is implemented (to see this, select ws_cr_st in view at the top left of the pane. When you are a few codelets in---pressing K jumps ahead 100 codelets---click on one of the codelet families and you will see details of the codelets of that family.

Thanks,
  Abhijit

**Comment by superMDguy** [comment]
I'm having some trouble figuring out exactly what's going on in the History class.  I get the overall concept, but some of the variable name abbreviations are hard for me to figure out.  Is that okay if I expand some of the variable names to improve readability?

**Comment by amahabal** [comment]
Go for it. If there are some portions particularly hard to grok, drop me a
line and I will spend some time making it more readable.

**Comment by superMDguy** [comment]
What's the difference between the `ObjectHistory` method and the `EventsForItem` method?

**Comment by amahabal** [comment]
The EventsForItems returns what events an object participated in (including its creation). ObjectHistory, by contrast, gives the ancestry: codelet X was created by codelet Y which was created when focusing on object Z which was created by codelet W, and so forth.

**Comment by superMDguy** [comment]
Earlier, you said you'd like to "make things navigable via clicking".  Currently, I've added a method to output the complete list of HID's, and some basic info about each.  Should I just make the HID clickable?  I felt like the way I currently have it set up has too many HID's to be able to easily navigate through.

**Comment by superMDguy** [comment]
Here's an example of the list I was talking about.  You can see that it'd be difficult to manage once it gets to a few hundred codelets.

> HID: 0	Object Type: CONTROLLER
> 		Class Name: SeqseeController
> 		Log Message: Created controller
> 
> HID: 1	Object Type: CODELET
> 		Class Name: CF:CF_ReadFromWS
> 		Log Message: Codelet CF_ReadFromWS Routine codelet
> 
> HID: 2	Object Type: CODELET
> 		Class Name: CF:CF_RemoveSpuriousRelations
> 		Log Message: Codelet CF_RemoveSpuriousRelations Routine codelet
> 
> HID: 3	Object Type: WS_GROUP
> 		Class Name: SAnchored
> 		Log Message: Initial creation
> 
> HID: 4	Object Type: WS_GROUP
> 		Class Name: SAnchored
> 		Log Message: Initial creation
> 
> HID: 5	Object Type: WS_GROUP
> 		Class Name: SAnchored
> 		Log Message: Initial creation
> 
> HID: 6	Object Type: WS_GROUP
> 		Class Name: SAnchored
> 		Log Message: Initial creation
> 
> HID: 7	Object Type: WS_GROUP
> 		Class Name: SAnchored
> 		Log Message: Initial creation

**Comment by amahabal** [comment]
I see. Okay, let me suggest an alternative design.

The goal is not to be able to see everything, but to see a few examples per
object type. To that end, in the new design, we can continue to have the
notebook pages for summary and counts. The summary page currently shows up
to 10 ids per object type. We can make these clickable. Clicking one would
change the notebook page to the third page, with that object's top details.

For the class that holds the UI for history, we can remember which item's
details are displayed on the  third page, so that refresh would work (we
can then see the action on something as the app runs normally).

I will like to add that the sort of duplication you point out was one of
the main reasons that I started on farg/apps/pyseqsee, and in the new
design there, there is far less mindless creation-destruction-recreation of
groups.

(You probably already are aware of this, but just checking: now, with
--history, the GUI window co-exists with the main window and is refreshed
as the run progresses in the main window)

Thanks,
  Abhijit

**Comment by superMDguy** [comment]
Aren't the numbers in the summary eid's?

**Comment by amahabal** [comment]
Yes, that's correct (they are the values of the dict returned by
GroupObjectsByClass). Currently, in the UI, they are not clickable.

**Comment by amahabal** [comment]
Heads up: I will change the history UI a bit today (perhaps along with the
change I mention above). I am meeting Doug this evening for a few
minutes,and wanted to show him some of this stuff. Apologies if you are in
the middle of a similar change (if you are, we will try to merge your work).

Regards,
  Abhijit

**Comment by superMDguy** [comment]
Ok, thanks for letting me know.  I'll commit what I have so far.

**Comment by amahabal** [comment]
Just saw your commit in dev. LGTM, I will merge it in tonight.

Thanks,
  Abhijit
